Mathlib/FieldTheory/SeparableDegree.lean

```diff
@@ -268,31 +249,548 @@ def embProdEmbOfIsAlgebraic [Algebra E K] [IsScalarTower F E K] (halg : Algebra.
+    simp only [RingEquiv.toRingHom_eq_coe, Multiset.mem_toFinset, mem_roots',
+      ne_eq, IsRoot.def] at h ⊢
+    use fun hf ↦ by simp [hf, Polynomial.map_zero, not_true, false_and] at h
+    replace h := congr_arg (frobeniusEquiv R p) h.2
+    have := eval₂_hom (p := g) (frobeniusEquiv R p).toRingHom x
+    simp only [RingEquiv.toRingHom_eq_coe, RingHom.coe_coe] at this
+    simpa only [← this, eval₂_map, RingEquiv.comp_symm, map_zero] using h⟩
+  let invFun : f.roots.toFinset → g.roots.toFinset := fun ⟨x, h⟩ ↦ ⟨(frobeniusEquiv R p).symm x, by
+    simp only [RingEquiv.toRingHom_eq_coe, Multiset.mem_toFinset, mem_roots',
+      ne_eq, IsRoot.def] at h ⊢
+    use (Polynomial.map_ne_zero_iff <| EquivLike.injective _).2 h.1
+    replace h := congr_arg (frobeniusEquiv R p).symm h.2
+    have := eval₂_hom (p := f) (frobeniusEquiv R p).symm.toRingHom x
+    simp only [RingEquiv.toRingHom_eq_coe, RingHom.coe_coe] at this
+    simpa only [← this, eval_map, map_zero] using h⟩
+  refine ⟨toFun, invFun, fun ⟨x, h⟩ ↦ ?_, fun ⟨x, h⟩ ↦ ?_⟩
+  · simp only [RingEquiv.toRingHom_eq_coe, frobeniusEquiv_apply,
+      frobeniusEquiv_symm_apply_frobenius]
+  · simp only [RingEquiv.toRingHom_eq_coe, frobeniusEquiv_apply,
+      frobenius_apply_frobeniusEquiv_symm]
+
+/-- If `f` is a polynomial over a perfect integral domain `R` of characteristic `p`, then there is
+a bijection from the set of roots of `Polynomial.expand R (p ^ n) f` to the set of roots of `f`.
+In fact it's given by `x ↦ x ^ (p ^ n)`, but we don't give a proof here. -/
+def rootsExpandPowEquivRoots
+    (R : Type u) [CommRing R] [IsDomain R]
+    (p : ℕ) [Fact p.Prime] [CharP R p] [PerfectRing R p] {f : R[X]} {n : ℕ} :
+    (expand R (p ^ n) f).roots.toFinset ≃ f.roots.toFinset := by
+  induction' n with n ih
+  · rw [pow_zero, expand_one]
+  · rw [pow_succ, ← expand_expand]
+    exact (rootsExpandEquivRoots R p).trans <| ih
+
+/-- If a field `F` is of exponential characteristic `q`, then `Polynomial.expand F (q ^ n) f`
+and `f` have the same separable degree. -/
+theorem natSepDegree_expand_eq_natSepDegree (q : ℕ) [hF : ExpChar F q] {n : ℕ} :
+    (expand F (q ^ n) f).natSepDegree = f.natSepDegree := by
+  cases' hF with _ _ hprime _
+  · simp only [one_pow, expand_one]
+  haveI := Fact.mk hprime
+  simpa only [natSepDegree_eq_of_isAlgClosed (AlgebraicClosure F), aroots_def, map_expand,
+    Fintype.card_coe] using Fintype.card_eq.2 ⟨rootsExpandPowEquivRoots q
+      (f := f.map (algebraMap F (AlgebraicClosure F))) (n := n)⟩
+
+variable {f} in
+/-- If a polynomial has separable contraction, then its separable degree is equal to the degree of
+the given separable contraction. -/
+theorem HasSeparableContraction.natSepDegree_eq
+    {q : ℕ} [ExpChar F q] (hf : f.HasSeparableContraction q) :
+    f.natSepDegree = hf.degree := by
+  have hf' := hf
+  obtain ⟨g, h1⟩ := hf'
+  rw [← IsSeparableContraction.degree_eq q hf g h1]
```

Comment:
> The API around `contraction` seems under-developed, e.g. there's no `isSeparableContraction_contraction` which is just `choose_spec`. You wouldn't need the rewrite if you add that lemma.

Let me check later.

Reply:
> The API around `contraction` seems under-developed, e.g. there's no `isSeparableContraction_contraction` which is just `choose_spec`. You wouldn't need the rewrite if you add that lemma.

Sorry, but I don't quite get what you mean. All I can think of is to extract one of the `rw` to `IsSeparableContraction.natSepDegree_eq`, and I only need one `rw` in each of the result. Is it OK?

Reply:
You can also do
```
  conv_lhs => rw [← h2, natSepDegree_expand_eq_natSepDegree]
  exact natSepDegree_eq_natDegree_of_separable _ h1
```
to avoid the `f` inside `hf` in the RHS being rewritten, causing a motive not type correct error.